File: itx-java/itx-impl/src/main/java/io/github/chehsunliu/itx/impl/repo/Jdbc.java
package io.github.chehsunliu.itx.impl.repo;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import javax.sql.DataSource;

public final class Jdbc {
    private Jdbc() {}

    @FunctionalInterface
    public interface ConnectionFn<T> {
        T apply(Connection conn) throws SQLException;
    }

    @FunctionalInterface
    public interface ConnectionVoidFn {
        void apply(Connection conn) throws SQLException;
    }

    @FunctionalInterface
    public interface ResultSetMapper<T> {
        T map(ResultSet rs) throws SQLException;
    }

    public static <T> T useConnection(DataSource ds, ConnectionFn<T> fn) {
        try (Connection conn = ds.getConnection()) {
            return fn.apply(conn);
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public static void useConnectionVoid(DataSource ds, ConnectionVoidFn fn) {
        useConnection(ds, conn -> {
            fn.apply(conn);
            return null;
        });
    }

    public static <T> T transactionally(DataSource ds, ConnectionFn<T> fn) {
        try (Connection conn = ds.getConnection()) {
            boolean originalAutoCommit = conn.getAutoCommit();
            conn.setAutoCommit(false);
            try {
                T result = fn.apply(conn);
                conn.commit();
                return result;
            } catch (Throwable t) {
                conn.rollback();
                throw t;
            } finally {
                conn.setAutoCommit(originalAutoCommit);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public static void bindAll(PreparedStatement ps, Object... args) throws SQLException {
        for (int i = 0; i < args.length; i++) {
            ps.setObject(i + 1, args[i]);
        }
    }

    public static <T> List<T> mapAll(ResultSet rs, ResultSetMapper<T> mapper) throws SQLException {
        List<T> out = new ArrayList<>();
        while (rs.next()) out.add(mapper.map(rs));
        return out;
    }

    public static <T> Optional<T> firstOrNull(ResultSet rs, ResultSetMapper<T> mapper) throws SQLException {
        return rs.next() ? Optional.of(mapper.map(rs)) : Optional.empty();
    }
}

File: itx-kt/itx-impl/src/main/kotlin/io/github/chehsunliu/itx/impl/repo/Jdbc.kt
package io.github.chehsunliu.itx.impl.repo

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import javax.sql.DataSource

internal suspend inline fun <T> DataSource.useConnection(crossinline block: (Connection) -> T): T =
    withContext(Dispatchers.IO) { connection.use { block(it) } }

internal suspend inline fun <T> DataSource.transactionally(crossinline block: (Connection) -> T): T =
    withContext(Dispatchers.IO) {
        connection.use { conn ->
            val originalAutoCommit = conn.autoCommit
            conn.autoCommit = false
            try {
                val result = block(conn)
                conn.commit()
                result
            } catch (t: Throwable) {
                conn.rollback()
                throw t
            } finally {
                conn.autoCommit = originalAutoCommit
            }
        }
    }

internal fun PreparedStatement.bindAll(vararg args: Any?) {
    args.forEachIndexed { i, value -> setObject(i + 1, value) }
}

internal fun <T> ResultSet.mapAll(transform: (ResultSet) -> T): List<T> {
    val out = mutableListOf<T>()
    while (next()) out.add(transform(this))
    return out
}

internal fun <T> ResultSet.firstOrNull(transform: (ResultSet) -> T): T? = if (next()) transform(this) else null
